src/gencost/waterfall.py

```diff
@@ -2256,3 +2267,480 @@ def valid_generation(df_):
+        )
+
+        # create a row for each year that needs to be filled in
+        split = pd.DataFrame([pd.Series(x) for x in missing_years.years_to_fill_in])
+        split.columns = [f"year_{x + 1}" for x in split.columns]
+
+        missing_years_clean = pd.concat([missing_years, split], axis=1)
+
+        missing_years_clean = (
+            missing_years_clean.melt(
+                id_vars=["plant_id_eia", "generator_id"],
+                value_vars=[
+                    "year_1",
+                    "year_2",
+                    "year_3",
+                    "year_4",
+                    "year_5",
+                    "year_6",
+                    "year_7",
+                    "year_8",
+                    "year_9",
+                    "year_10",
+                    "year_11",
+                    "year_12",
+                    "year_13",
+                    "year_14",
+                ],
+                value_name="year",
+            )
+            .query("year.notnull()")
+            .assign(
+                year=lambda x: x["year"].astype("Int64"), fuss=lambda x: "missing_years"
+            )
+            .merge(
+                self.get_gf923_by_generator(counterfactuals=True)
+                .groupby(
+                    [
+                        "plant_id_eia",
+                        "generator_id",
+                        pd.Grouper(key="report_date", freq="YS"),
+                        "prime_mover_code",
+                        "fuel_group",
+                    ]
+                )
+                .agg({"mmbtu": "sum", "net_mwh": "sum"})
+                .reset_index()  # keep latest prime fuel observation in gf 923 (largest single fuel)
+                # .query('report_date == "2020-01-01"')
+                # .query("mmbtu > 0 & net_mwh > 0") do we want to put prime/fuel of gens in GF
+                # reporting zeros?
+                .sort_values(
+                    by=["plant_id_eia", "generator_id", "report_date", "mmbtu"],
+                    ascending=True,
+                ).drop_duplicates(subset=["plant_id_eia", "generator_id"], keep="last")[
+                    [
+                        "plant_id_eia",
+                        "generator_id",
+                        "report_date",
+                        "prime_mover_code",
+                        "fuel_group",
+                    ]
+                ],
+                on=["plant_id_eia", "generator_id"],
+                how="left",
+            )
+            .assign(
+                prime_mover=lambda x: x.prime_mover_code.replace(
+                    FOSSIL_PRIME_MOVER_MAP
+                ),
+                age_in_current_year=lambda x: 1,
+            )
+            .merge(
+                df_860.sort_values(
+                    by=["plant_id_eia", "generator_id", "report_date"], ascending=True
+                ).drop_duplicates(
+                    subset=["plant_id_eia", "generator_id", "report_date"]
```

Comment:
I don't think you need the report_date here in drop_duplicates, getting rid of it should allow you to validate this merge as "m:1" which is a nice assurance that the right this is happening.

Reply:
fun fact discovered through this change: oldest existing generator opened in 1893